### PR TITLE
Use architecture independent code paths in Array

### PR DESCRIPTION
**Pull request by joncham**
We at Unity have [in-progress work towards a modern .NET ecosystem](https://discussions.unity.com/t/coreclr-and-net-modernization-unite-2024/1519272) enabled by adopting CoreCLR in lieu of Mono and updating IL2CPP to support the .NET BCL. I am investigating how architecture specific code paths can be reduced if not completely removed in the .NET BCL. I understand this architecture agnostic approach is not the direction of .NET in general, but I am hoping to validate how to best approach this without maintaining a completely separate fork.

The current BCL, more specifically `System.Private.Corlib`, has a non-trivial number of preprocessor checks for architecture/bit specific code. At Unity, our IL2CPP AOT compiler has historically consumed the Mono BCL which was architecture agnostic. This enabled a number of benefits for us when targeting multiple architectures such as reduced iteration time via a single managed linker step and transpile step to C++, and reduced deployment size via sharing our binary metadata file between architectures (produced during transpile).

In this preliminary PR, I include two commits with possible approaches that I wanted to validate before submitting larger PRs. In the first commit, native sized integers (`nint` & `nuint`) are used rather than a preprocessor check and `int` & `long` types. In the second commit, the `int` & `long` types remain in use but are switched via a `IntPtr.Size` check.

A few questions:
1. Any preference on either commit as a general approach when possible?
2. What metrics should be evaluated when making such changes? Are there sufficient performance and size tests run as part of the automated PR process?
3. In [more complex scenarios where the preprocessor checks are used](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs#L1920), does a new architecture independent target, e.g. `TARGET_ARCH_NEUTRAL`, make sense to introduce and use alongside the existing code paths? I would expect this path to contain both implementations with a runtime check similar to commit 2 in this PR.

Thanks for any feedback!

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Array.cs [hunk 1]
```diff
@@ -916,29 +916,23 @@ public static int BinarySearch(Array array, int index, int length, object? value
...
-#if TARGET_64BIT
-                            case CorElementType.ELEMENT_TYPE_U:
-#endif
                                 result = GenericBinarySearch<ulong>(array, adjustedIndex, length, value);
                                 break;
+                            case CorElementType.ELEMENT_TYPE_I:
+                                result = GenericBinarySearch<nint>(array, adjustedIndex, length, value);
+                                break;
+                            case CorElementType.ELEMENT_TYPE_U:
+                                result = GenericBinarySearch<nuint>(array, adjustedIndex, length, value);
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Array.cs:934]:
```suggestion
                                result = (IntPtr.Size == 4) ? GenericBinarySearch<int>(array, adjustedIndex, length, value) : GenericBinarySearch<long>(array, adjustedIndex, length, value);
                                break;
                            case CorElementType.ELEMENT_TYPE_U:
                                result = (IntPtr.Size == 4) ? GenericBinarySearch<uint>(array, adjustedIndex, length, value) : GenericBinarySearch<ulong>(array, adjustedIndex, length, value);
```
Same pattern as below - avoids nint/nuint generic instantiations.

**Comment by jkotas** [comment]
> Any preference on either commit as a general approach when possible?

I think the runtime check approach is better in this specific case. It avoids code size hit from extra generic instantiations. (The check will be turned into no-op via trimmer substitutions or by the compiler, so it is not really a runtime check.)

> What metrics should be evaluated when making such changes?

Perf (throughput, memory consumption, code size) and maintainability.

> Are there sufficient performance and size tests run as part of the automated PR process?

Our perf and size tests run in outer loop. They do not run as part of the PR. We typically depend on target perf tests for PR validation as necessary.

> In [more complex scenarios where the preprocessor checks are used](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs#L1920), does a new architecture independent target, e.g. TARGET_ARCH_NEUTRAL, make sense to introduce and use alongside the existing code paths

I think the arch-neutral config will be leaving some perf on the table. A good example is: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs . I do not think there is a good way to make the arch-neutral variant that is perf-neutral. 

New `TARGET_ARCH_NEUTRAL` may help with readability of the code in some situations.

cc @stephentoub

**Comment by EgorBo** [comment]
> The check will be turned into no-op via trimmer substitutions or by the compiler, so it is not really a runtime check

ILLink still struggles with some, e.g. https://github.com/dotnet/runtime/issues/83169 (still a constant in the end, but affects inlining decisions/dll size).

File: src/libraries/System.Private.CoreLib/src/System/Array.cs (hunk 1 above)
Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Array.cs:934]: Could you please update all `#if TARGET_32/64BIT` in Array.cs to this pattern? We can merge this PR after that.

File: src/libraries/System.Private.CoreLib/src/System/Array.cs [hunk 2]
```diff
@@ -916,29 +916,23 @@ public static int BinarySearch(Array array, int index, int length, object? value
...
                                 break;
                             case CorElementType.ELEMENT_TYPE_U8:
-#if TARGET_64BIT
-                            case CorElementType.ELEMENT_TYPE_U:
-#endif
                                 result = GenericBinarySearch<ulong>(array, adjustedIndex, length, value);
                                 break;
+                            case CorElementType.ELEMENT_TYPE_I:
+                                result = (IntPtr.Size == 4) ? GenericBinarySearch<int>(array, adjustedIndex, length, value) : GenericBinarySearch<long>(array, adjustedIndex, length, value);
+                                break;
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Array.cs:932]:
```suggestion
                                if (IntPtr.Size == 4)
                                    goto case CorElementType.ELEMENT_TYPE_I4;
                                goto case CorElementType.ELEMENT_TYPE_I8;
```

File: src/libraries/System.Private.CoreLib/src/System/Array.cs [hunk 3]
```diff
@@ -916,29 +916,23 @@ public static int BinarySearch(Array array, int index, int length, object? value
...
+                            case CorElementType.ELEMENT_TYPE_U:
+                                result = (IntPtr.Size == 4) ? GenericBinarySearch<uint>(array, adjustedIndex, length, value) : GenericBinarySearch<ulong>(array, adjustedIndex, length, value);
+                                break;
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Array.cs:935]:
```suggestion
                                if (IntPtr.Size == 4)
                                    goto case CorElementType.ELEMENT_TYPE_U4;
                                goto case CorElementType.ELEMENT_TYPE_U8;
```

File: src/libraries/System.Private.CoreLib/src/System/Array.cs [hunk 4]
```diff
@@ -1424,19 +1418,15 @@ public static int IndexOf(Array array, object? value, int startIndex, int count)
...
                         case CorElementType.ELEMENT_TYPE_I8:
                         case CorElementType.ELEMENT_TYPE_U8:
-#if TARGET_64BIT
+                            result = GenericIndexOf<long>(array, value, adjustedIndex, count);
+                            break;
                         case CorElementType.ELEMENT_TYPE_I:
                         case CorElementType.ELEMENT_TYPE_U:
-#endif
-                            result = GenericIndexOf<long>(array, value, adjustedIndex, count);
+                            result = (IntPtr.Size == 4) ? GenericIndexOf<int>(array, value, adjustedIndex, count) : GenericIndexOf<long>(array, value, adjustedIndex, count);
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Array.cs:1429]:
```suggestion
                            if (IntPtr.Size == 4)
                                goto case CorElementType.ELEMENT_TYPE_I4;
                            goto case CorElementType.ELEMENT_TYPE_I8;
```

File: src/libraries/System.Private.CoreLib/src/System/Array.cs [hunk 5]
```diff
@@ -1424,19 +1418,15 @@ public static int IndexOf(Array array, object? value, int startIndex, int count)
...
                             break;
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Array.cs:1430]:
```suggestion
```

File: src/libraries/System.Private.CoreLib/src/System/Array.cs [hunk 6]
```diff
@@ -1654,19 +1644,15 @@ public static int LastIndexOf(Array array, object? value, int startIndex, int co
...
                         case CorElementType.ELEMENT_TYPE_I8:
                         case CorElementType.ELEMENT_TYPE_U8:
-#if TARGET_64BIT
+                            result = GenericLastIndexOf<long>(array, value, adjustedIndex, count);
+                            break;
                         case CorElementType.ELEMENT_TYPE_I:
                         case CorElementType.ELEMENT_TYPE_U:
-#endif
-                            result = GenericLastIndexOf<long>(array, value, adjustedIndex, count);
+                            result = (IntPtr.Size == 4) ? GenericLastIndexOf<int>(array, value, adjustedIndex, count) : GenericLastIndexOf<long>(array, value, adjustedIndex, count);
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Array.cs:1655]:
```suggestion
                            if (IntPtr.Size == 4)
                                goto case CorElementType.ELEMENT_TYPE_I4;
                            goto case CorElementType.ELEMENT_TYPE_I8;
```

File: src/libraries/System.Private.CoreLib/src/System/Array.cs [hunk 7]
```diff
@@ -1654,19 +1644,15 @@ public static int LastIndexOf(Array array, object? value, int startIndex, int co
...
                             break;
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Array.cs:1656]:
```suggestion
```

File: src/libraries/System.Private.CoreLib/src/System/Array.cs [hunk 8]
```diff
@@ -1863,22 +1849,21 @@ public static void Reverse(Array array, int index, int length)
...
                 case CorElementType.ELEMENT_TYPE_R8:
                     UnsafeArrayAsSpan<long>(array, adjustedIndex, length).Reverse();
                     return;
+                case CorElementType.ELEMENT_TYPE_I:
+                case CorElementType.ELEMENT_TYPE_U:
+                    if (IntPtr.Size == 4)
+                        UnsafeArrayAsSpan<int>(array, adjustedIndex, length).Reverse();
+                    else
+                        UnsafeArrayAsSpan<long>(array, adjustedIndex, length).Reverse();
+                    return;
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Array.cs:1866]:
```suggestion
                    if (IntPtr.Size == 4)
                        goto case CorElementType.ELEMENT_TYPE_I4;
                    goto case CorElementType.ELEMENT_TYPE_I8;
```

File: src/libraries/System.Private.CoreLib/src/System/Array.cs [hunk 9]
```diff
@@ -2069,29 +2054,29 @@ public static void Sort(Array keys, Array? items, int index, int length, ICompar
...
+                                GenericSort<int>(keys, items, adjustedIndex, length);
+                            else
+                                GenericSort<long>(keys, items, adjustedIndex, length);
+                            return;
+                        case CorElementType.ELEMENT_TYPE_U:
+                            if (IntPtr.Size == 4)
+                                GenericSort<uint>(keys, items, adjustedIndex, length);
+                            else
+                                GenericSort<ulong>(keys, items, adjustedIndex, length);
+                            return;
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Array.cs:2079]:
```suggestion
                                goto case CorElementType.ELEMENT_TYPE_I4;
                            goto case CorElementType.ELEMENT_TYPE_I8;
                        case CorElementType.ELEMENT_TYPE_U:
                            if (IntPtr.Size == 4)
                                goto case CorElementType.ELEMENT_TYPE_U4;
                            goto case CorElementType.ELEMENT_TYPE_U8;
```

**Review by jkotas — COMMENTED** [review]
Looking at some of the long lines, I have realized that we can just use `goto case`

**Review by jkotas — APPROVED** [review]

**Comment by am11** [comment]
> TARGET_ARCH_NEUTRAL

Does this aim to cover only differences in bitness (i.e., 32-bit vs. 64-bit), or would it also encompass architecture types (e.g., x86/x64 vs. ARM/ARM64)?

**Comment by joncham** [comment]
> > TARGET_ARCH_NEUTRAL
> 
> Does this aim to cover only differences in bitness (i.e., 32-bit vs. 64-bit), or would it also encompass architecture types (e.g., x86/x64 vs. ARM/ARM64)?

I envision it cover architecture types as well. My intent is to minimize any bit/architecture specific paths within the IL.

**Pull request merged**